src/Bridge/Symfony/Bundle/DependencyInjection/Compiler/ValidationGroupsGeneratorFilterPass.php

```diff
@@ -0,0 +1,42 @@
+<?php
+
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Bridge\Symfony\Bundle\DependencyInjection\Compiler;
+
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Reference;
+
+/**
+ * Registers validation groups generators.
+ *
+ * @internal
+ *
+ * @author Tomas Norkūnas <norkunas.tom@gmail.com>
+ */
+final class ValidationGroupsGeneratorFilterPass implements CompilerPassInterface
```

Comment:
Can't this class be removed in favor of `!tagged`? https://symfony.com/doc/current/service_container/tags.html#reference-tagged-services

Reply:
I'm not sure, because `ServiceLocator` holds the index to the service so this compiler pass maintains the indexes, otherwise we should use `index_by` which was introduced in Sf 4.3 and then the generators should define the method to get the index. Prove me wrong!

Reply:
Also service locator throws this when trying to inject !tagged:
`Invalid definition for service "app_service_locator": an array of references is expected as first argument when the "container.service_locator" tag is set.`

Reply:
See https://github.com/api-platform/core/pull/3346/files#r383869931

Let's only do this in API Platform 3.0 - just stick to the full container for now.